ui/src/components/UploadStory.tsx

```diff
@@ -0,0 +1,855 @@
+});
+
+function useResetCache(data: any) {
+  const ref = React.useRef<VariableSizeList>(null);
+  React.useEffect(() => {
+    if (ref.current != null) {
+      ref.current.resetAfterIndex(0, true);
+    }
+  }, [data]);
+  return ref;
+}
+
+// Adapter for react-window
+const ListboxComponent = React.forwardRef<HTMLDivElement>(
+  function ListboxComponent(props, ref) {
+    const { children, ...other } = props;
+    const itemData = React.Children.toArray(children);
+    const theme = useTheme();
+    const smUp = useMediaQuery(theme.breakpoints.up("sm"), { noSsr: true });
+    const itemCount = itemData.length;
+    const itemSize = smUp ? 36 : 48;
+
+    const getChildSize = (child: React.ReactNode) => {
+      if (React.isValidElement(child) && child.type === ListSubheader) {
+        return 48;
+      }
+
+      return itemSize;
+    };
+
+    const getHeight = () => {
+      if (itemCount > 8) {
+        return 8 * itemSize;
+      }
+      return itemData.map(getChildSize).reduce((a, b) => a + b, 0);
+    };
+
+    const gridRef = useResetCache(itemCount);
+
+    return (
+      <div ref={ref}>
+        <OuterElementContext.Provider value={other}>
+          <VariableSizeList
+            itemData={itemData}
+            height={getHeight() + 2 * LISTBOX_PADDING}
+            width="100%"
+            ref={gridRef}
+            outerElementType={OuterElementType}
+            innerElementType="ul"
+            itemSize={(index) => getChildSize(itemData[index])}
+            overscanCount={5}
+            itemCount={itemCount}
+          >
+            {renderRow}
+          </VariableSizeList>
+        </OuterElementContext.Provider>
+      </div>
+    );
+  }
+);
+
+const renderGroup = (params: AutocompleteRenderGroupParams) => [
+  <ListSubheader key={params.key} component="div">
+    {params.group}
+  </ListSubheader>,
+  params.children,
+];
+
+interface InputProps {
+  onKeyDown: (
+    event: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>
+  ) => void;
+}
+
+interface TagParameters {
+  inputProps: InputProps;
+}
+
+interface StoryProps {
+  id: number;
+  currentStory: Story;
+  bio: string;
+}
+
+export const UploadStory: React.FC<StoryProps> = ({
+  id,
+  currentStory,
+  bio,
+}: StoryProps) => {
+  const { data: tagOptions, error } = useSWR<string[]>("/api/tags");
+  const [tagArray, setTagArrayValues] = useState(currentStory.tags);
+  const [authorCountry, setAuthorCountry] = useState(
+    currentStory.author_country
+  );
+  const [currentCity, setCurrentCity] = useState(
+    currentStory.current_city.toUpperCase()
+  );
+  const [autocompleteAuthorCountry, setAutocompleteAuthorCountry] = useState(
+    currentStory.author_country
+  );
+  const [addedCountry, setAddedCountry] = useState("");
+  const [newImage, setNewImage] = useState(currentStory.image_url);
+  const [disabled, setDisabled] = useState(false);
+  //handleSubmit component states
+  const [dialogOpenState, setDialogOpenState] = useState(false);
```

Comment:
multiple use states is a code smell. Let's refactor this into `useReducer` (see https://github.com/uwblueprint/shoe-project/pull/267/files#diff-5c3739858bec843543a6cf139cc3b2e3febe6fe751e55dbb2c20daf08f8d76b2R1 for an example)